### PR TITLE
general: fix error 400 bad request

### PR DESCRIPTION
**Pull request by Garfield-fr**
* Closes rero/rero-ils#2705.
* Removes localeStorage user data.
* Updates guards that used the user localeStorage.
* Improves somes guards.
* Adds some tests for guards.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

**Review by jma — CHANGES_REQUESTED** [review]
Some code still need to be removed such as: https://github.com/rero/rero-ils-ui/blob/972884d72ec8053be05b529f0a3a1408ef8be955/projects/admin/src/app/service/app-router-event.service.ts#L28 or `STORAGE_KEY`.

File: projects/admin/src/app/guard/can-access.guard.ts [hunk 1]
```diff
@@ -15,45 +15,53 @@
...
-      } else {
-        this._router.navigate(['/errors/400'], { skipLocationChange: true });
-      }
+    if (params && params.type && params.pid) {
+      return this._recordService.getRecord(params.type, params.pid)
+        .pipe(map((record: any) => {
+          const user = this._userService.user;
+          if (user.currentOrganisation !== extractIdOnRef(record.metadata.organisation.$ref)) {
+            /** Access forbidden */
+            this._router.navigate(['/errors/403'], { skipLocationChange: true });
```
Comment (zannkukai) [inline projects/admin/src/app/guard/can-access.guard.ts:57]: missing `return false`;

File: projects/admin/src/app/guard/can-add-local-fields.guard.ts [hunk 2]
```diff
@@ -54,35 +54,24 @@ export class CanAddLocalFieldsGuard implements CanActivate {
...
-        }));
-      }
-      this._router.navigate(['/errors/401'], { skipLocationChange: true });
+      const type = this._translateType(params.type);
+      return this._localFieldsApiService.getByResourceTypeAndResourcePidAndOrganisationId(
+        type,
+        params.ref,
+        this._userService.user.currentOrganisation
+      ).pipe(map(record => {
+        return record.metadata ? false : true;
```
Comment (zannkukai) [inline projects/admin/src/app/guard/can-add-local-fields.guard.ts:69]: cosmetic : I learned the `!!` operator yesterday to convert anything to boolean : https://brianflove.com/2014-09-02/whats-the-double-exclamation-mark-for-in-javascript/
So you should use  `return !!(record.metadata);` (to be tested)

File: projects/admin/src/app/guard/can-update.guard.ts [hunk 3]
```diff
@@ -33,23 +34,24 @@ export class CanUpdateGuard implements CanActivate {
...
-    this._permissionService.getPermission(next.params.type, next.params.pid).subscribe(
-      (permission: RecordPermissions) => {
+  canActivate(next: ActivatedRouteSnapshot): Observable<boolean> {
+    return this._permissionService.getPermission(next.params.type, next.params.pid).pipe(
+      map((permission: RecordPermissions) => {
         if (!permission.update.can) {
           this._router.navigate(['/errors/403'], { skipLocationChange: true });
+          return false;
+        } else {
+          return true;
```
Comment (zannkukai) [inline projects/admin/src/app/guard/can-update.guard.ts:52]: you could remove the `else` statement to return `true` by default

File: projects/admin/src/app/guard/library.guard.ts [hunk 4]
```diff
@@ -16,50 +16,52 @@
...
-    combineLatest([
-      this.getOwningLibrary$(next),
-      this._userService.loaded$
-    ]).subscribe(
-      ([owningLibrary, userMetadata ]) => {
+  canActivate(next: ActivatedRouteSnapshot): Observable<boolean> {
+    return this.getOwningLibrary$(next).pipe(
+      map((owningLibrary: string) => {
         if (owningLibrary !== this._userService.user.currentLibrary) {
           this._router.navigate(['/errors/403'], { skipLocationChange: true });
```
Comment (zannkukai) [inline projects/admin/src/app/guard/library.guard.ts:51]: missing the `return false`

**Review by zannkukai — CHANGES_REQUESTED** [review]

**Review by benerken — APPROVED** [review]

**Review by zannkukai — APPROVED** [review]

**Review by jma — APPROVED** [review]

**Pull request merged**